### ISSUE TITLE
support history expansion (e.g., bash's !!, bang bang, and !$, bang dollar, tokens)

### ISSUE THREAD

**Issue by cometpeak** [fish-shell/fish-shell#288]
It appears that the double-bang does not work and is somehow interpreted as a program which sudo unsuccessfully tries to locate. This occurs on Ubuntu 12.04 LTS with fish-shell compiled from source.
## Steps to repro:
1. Execute any command:
   
   $ ls
2. Execute the previous command as root:
   
   $ sudo !!
   sudo: !!: command not found
## Expected behavior:

Command `sudo !!` executes previously executed command (`ls`) as root.

I also checked that the issue isn't caused by `sudo` by following repro steps in bash, which behaved as expected.

Also, note that an earlier version of fish-shell on Mac OS X 10.7 did not have this problem and `sudo !!` worked as expected.

**Comment by gillham** [comment]
The argument of "echo fish is great!!" being better than using '!!' for history is a bit weak, but I understand the point.  Seems like a simple "set shell_history advanced" or something could allow it for people willing to risk "echo fish is great!!" getting an error message.

Nevertheless, my use case of '!!' is similar but could be avoided with some feature enhancement perhaps.

For example on my mac with bash:
    $ brew update
    $ brew outdated
    $ brew fetch `!!`

So I somewhat regularly use the output of the previous command for a new command.
With fish I can do this:
    $ brew fetch (brew outdated)
Now at some point perhaps that will be a suggestion I can use, but what I really need is to be able to search the history within the () block.
e.g.:
    $ brew fetch (bre<up-arrow>)

Unfortunately even though fish is fairly proud of using () for sub-commands instead of "confusing" backticks and such, it doesn't really treat them as such until you hit enter.  At least that is what I see.

If fish would allow a history search after the first '(' (aka a subcommand) I think it would be easier to handle not having ! history manipulation.

**Comment by bucaran** [comment]
-1

**Comment by ridiculousfish** [comment]
If I understand the suggestion, it's that when you press up arrow, it should perform history replacement only within the innermost subcommand instead of the entire command line.

**Comment by SirCmpwn** [comment]
:-1: 

I still support simply implementing bash-style `!!`.

**Comment by mrak** [comment]
I'd prefer :up_arrow:, Ctrl+p and command history search working in subcommand positions over introducing !!

The entire history API of other shells is confusing enough that people usually only remember !! from it.

**Comment by gillham** [comment]
Correct. That would be convenient and fairly easily replace the '!!' Or !keyword use.

**Comment by ElijahLynn** [comment]
Really like the idea of `!!` being an abbreviation.

**Comment by mdsib** [comment]
It looks like abbreviations in the most recent nightly builds are only allowing abbreviations in the command position, hopefully it'll be updated to be able to be used as arguments for cases like this one.

**Comment by bucaran** [comment]
This feature is overrated, here is an alternative:

```fish
$ rm -rf $dir
```

Now, execute the previous command as root:

```fish
$ sudo rm -rf $dir
```

**Comment by ElijahLynn** [comment]
@bucaran I am not following exactly. The `!!` is for speed, how do I match the speed of `!!` with `cmd`?

Do you mean make a variable called `$cmd` that is the last command? It isn't a native variable.

Thanks

**Comment by ElijahLynn** [comment]
@mdsib Yeah, I think I remember that there was a discussion in another issue about not allowing abbreviations in the argument position. At the time it was preferred to not implement that. This would definitely change that.

**Comment by SirCmpwn** [comment]
I think the takeaway from this thread is that it's not happening any time soon, and I should unsubscribe from this Github issue and switch to zsh. Adios!

**Comment by bucaran** [comment]
@ElijahLynn I am just saying that `!!` is an overrated feature. I don't need it, and I hope it never makes it to `fish`/

**Comment by SirCmpwn** [comment]
Not because `!!` isn't here, but because a Github issue for a fairly simple feature has been open for 3 years without being implemented due to philosophical BS.

**Comment by bucaran** [comment]
The fact there is an issue open is of course an indication some people want the feature, but the fact it's not implemented also shows you some other (more for sure) do not agree. That philosophical BS is what makes fish different from other shells.

Is `!!` really such a big deal?

**Comment by gillham** [comment]
It is funny to see people trying to convince others that a feature isn't a big deal.  Sure, you don't want it, so you don't care.. It is like you're trying convince people not to like green because you think blue is better.
Nevertheless since a "simple" feature can't seem to make it in, perhaps someone should just fork fish if they like the existing features but would like to add a few more.  Seems like that would get somewhere quicker.

I made ~/bin/!! that looks like this:
#!/bin/fish
eval $history[1]

For the short term it helps me workaround the missing feature and my tendency to type !!.
I also made ~/bin/!vi:
#!/bin/fish
for f in $history
	echo $f | grep '^vi'
	if test "$status" = "0"
		eval $f
		exit $status
	end
end

Not ideal, but enough that my shell tends to do what I expect more often now.

**Comment by bucaran** [comment]
https://github.com/fish-shell/fish-shell/network#fork-destination-box

**Comment by ridiculousfish** [comment]
@SirCmpwn  Is it not obvious to me that `!!` is a simple feature. For example, bash supports this:

    ls some_long_file.txt
    !!:s/ls -l/cat/

Should this use of !! be supported or not? Users who say "fish should support !!" may have different ideas of what that means.

You feel strongly about this, so I encourage you (or anyone) to author a proposal for how !! would work in fish. What syntax would be supported exactly? How would this expansion relate to other types of expansion, like variable or process expansion? How would argument splitting occur in the expanded string? Could you use escapes to prevent it?

This is a sincere request. A feature is not defined by a single use case; we need a concrete proposal to discuss.

If you say "I want this case to work, and I don't care how," then that's fine and valid. But someone has to work out the details before anything can be implemented.

**Comment by ridiculousfish** [comment]
In the near term, one concrete plan is to implement global aliases and then make !! into one. An alternative is to merge the docopt work, give a function signature for sudo, and then make !! into a regular alias. Which direction we go hinges on whether we think !! should expand everywhere, or just in command position.

**Comment by gillham** [comment]
@ridiculousfish You're absolutely right that a '!!' feature request needs to be fairly detailed.  I've never used the particular example you gave as I always do something like this:
$ ls -l some_long_file.txt
$ ^ls-l^cat

Obviously this doesn't work either.  That would be separate feature request for me, that I haven't bothered with since I only started testing fish out recently.

"fish: Encountered redirection when expecting a command name. Fish does not allow a redirection operation before a command."

Seems like it would be easy to treat '^' in the command position as a substitution since it isn't conflicting with existing functionality.  But that is a separate issue, just mentioning it since you're right that there should be a clear definition of what '!!' (or just history substitution in general) means.

Here are a couple examples I use:
```
$ !!
$ some command `!!`
$ !keyword 
$ !123
```

**Comment by ridiculousfish** [comment]
Thank you @gillham , information like that is very helpful when designing a feature.

**Comment by zanchey** [comment]
I also note that @nyarly has a clever workaround for `!!` and `!$` here: https://github.com/fish-shell/fish-shell/wiki/Bash-Refugees

**Comment by anordal** [comment]
Better fix #5:
Fish is so history driven, that when having forgot sudo or otherwise polluted your history with a non-working command, it may actually be worth your time (in the long run) to manually erase it from history. Yes, that's open a new terminal, exec bash and edit ~/.config/fish/fish_history. Fish has led me onto the wrong command time and time again, when I have skimped on this.

As for @nyarly's `mkdir …; cd !$` habit, I'm doing `mkcd …`:
```
function mkcd --description 'mkdir and cd'
        mkdir $argv[1]
        and cd $argv[1]
end
```

**Comment by ridiculousfish** [comment]
Note that erasing from history doesn't require bash shenanigans:

    history --delete --prefix some_command

`fish_config history` also lets you do it by point-n-click.

**Comment by AerysBat** [comment]
This functionality can be replicated quite well within Fish using https://github.com/fish-shell/fish-shell/wiki/Bash-Style-History-Substitution-(!!-and-!$)  -- perhaps this issue should be closed?

**Comment by geoff-codes** [comment]
:-1:  :-1: Unless a full bash compatibility mode is implemented.

**Comment by geoff-codes** [comment]
I think this is a slippery slope. If this is implemented in fish, what about single !s, and what about ksh history features? What about $()? Fish is a csh-style shell. If people can't/won't adapt, there's zsh...

**Comment by tetra-archos** [comment]
What I have gleamed from this discussion as a tl;dr for people who find this. (over 3 years in)
1. The team does not want to develop the feature, because they don't use it.
2. There are a lot of people who want the feature, because they use it.
3. The team seems unwilling to implement the feature as it exists but is willing to make something 
    that works similarly if we give them enough data.
4. People either accept not having the feature, implement a workaround that mostly works, or move 
    on to another project.

I am one of the latter, while I hope this get's resolved I cannot reasonably use fish and add that workaround with the number of machines I use without a considerable amount of effort (I need different config setups in different machines), and I use !! enough where it is a requirement for me. So I will be following this discussion in the hopes that one day fish will meet my use case.

**Comment by ridiculousfish** [comment]
@tetra-archos Nobody has put forth a concrete proposal yet, so we don't have a firm grasp of what "the feature" is. Since you obviously feel strongly about this, if you were so inclined, it would be a big help if you were to suggest what the behavior ought to be, as a starting point. Failing that, it would be useful to at least list the history substitution forms you use.

Note that the full bash history substitution behavior is very complex, and also configurable, so "just do what bash does" is a non-starter.

**Comment by faho** [comment]
>Nobody has put forth a concrete proposal yet, so we don't have a firm grasp of what "the feature" is.

@tetra-archos: This is absolutely critical. So far, people have just stated "We want "!!"!!" and maybe "!$" (which is "insert last argument of the last command"). But there's a lot of details that need to be banged out if it (or a substitute) is to be implemented.

For example:

- Which part of the history substitutions do you want - is it just "insert the last command here" ("!!") and "insert the last command's last argument here" ("!$"), or all the other stuff like "!!:s/string1/string2" (I'd rather add a general "replace" mode like emacs' "query-replace" or vim's ":s/" so you can also do it _before_ executing the wrong command)? How about "insert the fifth command in my history" ("!5") - which I never quite got - or "insert the fifth-last command" ("!-5")? How about "[word designators](https://www.gnu.org/software/bash/manual/bashref.html#Word-Designators)" and "[modifiers](https://www.gnu.org/software/bash/manual/bashref.html#Modifiers)"?

- Do you want the command to execute right away (i.e. you enter `sudo !!`, press enter and the entire thing executes right away) or would it be okay to have it inserted first so you can double-check - this is like bash with the "histverify" option (I'd lean towards the latter)?

- Does it need to be "!!" and "!$" or can it be used differently , i.e. is muscle-memory your problem? If that's the case, though, we're never gonna be 100%, and history substitution _is_ a rather small part of bash

If it's just muscle-memory, you can also put the following in your fish_user_key_bindings:

```fish
bind '!!' 'commandline -i "$history[1]"'
bind '!$' 'history-token-search-backward'
```

(the issue this has, why we probably shouldn't take it in the default bindings, is that it then waits after you've entered "!" to check for a second character)

**Comment by andreaskern** [comment]
I think that `!!` is bad practice

The command `sudo !!` does not tell anyone what the shell will actually execute.
if i want execute the **`p`**revious command with `sudo` at the '**`a`**nfang' i type 

    ^p ^a sudo

I suggest adding a message to tell people to use the line editing commands ^p and ^a instead of using the history function.

    !!: command not found, try 'Ctrl-p Ctrl-a' to prefix the previous command with 'sudo' or any other  prefix

**Comment by pickfire** [comment]
I hate `!!` too. An easier method for `fish_vi_mode` as an alternative to `^p ^a`:

    ^v k I

**Comment by tetra-archos** [comment]
Because you don't like a way of doing something doesn't make it less valid.

**Comment by mrak** [comment]
The reverse is true, too.

Both sides of this issue have legitimate needs and concerns. I think the main holdup is finding the _right_ way to do it instead of the quick and dirty way

**Comment by tetra-archos** [comment]
I agree and I have nothing against the other ways to do it, they just don't
fit my (and obviously other people's) needs.

**Comment by Ahti** [comment]
To reiterate the approaches mentioned in the early discussion and to expand on them a little: I have found keybindings to work better for me than "!!".

I've written this function which can be bound to some key (^S for me, `bind \cs 'prepend_command sudo'`). If the command line has content, it prepends sudo. If there is no content, it prepends sudo to the last item in the history. So `sudo !!<Enter>` now becomes `^S<Enter>`.

```fish
function prepend_command
  set -l prepend $argv[1]
  if test -z "$prepend"
    echo "prepend_command needs one argument."
    return 1
  end

  set -l cmd (commandline)
  if test -z "$cmd"
    commandline -r $history[1]
  end

  set -l old_cursor (commandline -C)
  commandline -C 0
  commandline -i "$prepend "
  commandline -C (math $old_cursor + (echo $prepend | wc -c))
end
```

Some people said configuring something like this is too complicated to do. A possible solution might be to include something functionally similar to my function in fish so configuring this becomes a one-liner at least.

Of course, this doesn't answer other demands like `some_command --with-subshell-param (!!)`.

**Comment by ElijahLynn** [comment]
@Ahti That is actually badass, just got it working, looks like it will be a treat to use, will report back in a while!

**Comment by bucaran** [comment]
@Ahti A plugin that add those bindings in [Fisherman](http://fisherman.sh) would be sweet.

**Comment by ElijahLynn** [comment]
Yeah, so far this beats `sudo !!`. Being halfway through a command and realizing it should be `sudo` happens to me a lot. `^s`, instant `sudo` prefix!

Personally I never used the `!!` in any context other than `sudo !!` so that doesn't bother me.

**Comment by ElijahLynn** [comment]
Coincidentally I just tried out fish_vi_mode and really like it. However, in Vi mode `^s` doesn't appear to bind anymore. Suggestions?

**Comment by ElijahLynn** [comment]
Okay, I had the key binding in config.fish. I found #2254 and then found https://github.com/faho/fish-shell/commit/f7e8ad632a388d5f1c5eb4cfc33f3cdb1f7fd831 and then found http://stackoverflow.com/a/16675092/292408 which confirmed I should place my bindings in a fish_user_key_bindings function. 

That worked and now when I am in Vi command mode `[N]` the prefix works as expected.

**Comment by zanchey** [comment]
We already have Alt-P to run the current pipeline in the pager; I wonder whether we should add Alt-S as well?

**Comment by hubbcaps** [comment]
I think @Ahti's is a pretty elegant solution to what everyone has been looking for. Direct to a keybind with no worry about a random expansion or weird edge cases. It simply places text for you to use or cancel out as you wish.

**Comment by orzechow** [comment]
I like both @nyarly's substitution and @Ahti's key binding solution.
Personally, I use `!!` only in the context of `sudo !!`

**Comment by geoff-codes** [comment]
Oy! This thread is almost four years old now. I recently bought ǃǃ.com and ǃǃ.net (which Github won't let me hyperlink, apparently, so copy-and-paste) if anyone wants them... taking into account that LATIN LETTER RETROFLEX CLICK is a different code point than EXCLAMATION MARK.

**Comment by Timopheym** [comment]
Hey! That's cool! @geoff-codes How did you do that?)

**Comment by geoff-codes** [comment]
They've recently started allowing .com and .net "internationalized" domain names.
<img width="922" alt="screen shot 2016-05-11 at 4 53 48 am" src="https://cloud.githubusercontent.com/assets/3162811/15180103/70540bf2-1734-11e6-8ee7-5c47309efbd6.png">

Support for it is still pretty buggy at this time, though.

**Comment by bucaran** [comment]
I didn't know you could register !!.com, that's crazy.

The lesson learned from this issue: if you ever write a Unix shell, the only feature you should make sure to implement is sudo !!, because it seems people can't live without it.

**Comment by geoff-codes** [comment]
@bucaran Yes, indeed, that seems to be the case 😆.

**Comment by floam** [comment]
International Domain Names - IDNs

On May 11, 2016 at 4:44:58 AM, Timophey (notifications@github.com) wrote:

Hey! That's cool! @geoff-codes <https://github.com/geoff-codes> How did you
do that?)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/fish-shell/fish-shell/issues/288#issuecomment-218435761>

**Comment by andreaskern** [comment]
In the bash for Windows presentation the guy presenting typed something
like this:

$> echo hello Windows!!

And I hope that that you can all guess what happened.

**Comment by base698** [comment]
So, I should switch to bash?

**Comment by nyarly** [comment]
@base698 - I wound up putting out a fisherman plugin to support the two most common use cases of history substitution: !! for the whole last command and !$ for the last argument. Honestly, I've mostly broken my own habits because fish's command line editing are so good. Still, I really like being able to 
```
> which tool
> cat (!!)
to examine a script for instance 

You can install the plugin with `fisher install nyarly/fish-bang-bang` once you've got fisherman installed.

**Comment by andreaskern** [comment]
for the sake of completeness (and giving a correct example)
here is a screenshot of the youtube video I mentioned, this is one of the great features of the '!!' pattern which is now avaible on Windows too. 
![bangpatternbash](https://cloud.githubusercontent.com/assets/739277/15965344/900500f4-2f1d-11e6-80a6-7d1eb207f100.png) 

https://www.youtube.com/watch?v=2dB0igTfhfg&feature=youtu.be&t=760

He wants to execute `git commit -a -m "Rawr!!"`

results in `git commit -a -m "Rawrclear"`

**Comment by pjrt** [comment]
As much as I love fish, !! is just too useful. Not sure `sudo !!` but also as a way to pass previous info to new commands. For example:

`find . -name something | grep "pattern"` then I would play with the pattern until I get what I want. 
Then I would pass it to something like so (someCommand `!!`). Currently, in fish, this requires too many steps. The workarounds don't work well enough for me. 

zsh expands !! so that you can edit it. So gonna have to go back to zsh.

**Comment by nyarly** [comment]
If you're using fisherman (which you very much should be) you can install `nyarly/fish-bang-bang` and get anywhere-in-the-command expansion of `!!` and `!$`

**Comment by pjrt** [comment]
@nyarly Yeah, I tried that. But I don't like how it expands immediately after the second `!` is typed. It felt awkward. zsh feels better because it expands after you enter a space or hit enter.

**Comment by Bahaa-Addin** [comment]
[Solved] Here:
https://blog.hospodarets.com/fish-shell-the-missing-config

**Comment by pgan002** [comment]
I'm so glad this was *not* implemented! It violates Fish's [Law of orthogonality](http://fishshell.com/docs/current/design.html) and Law of user focus. Bash's history replacement requires too much thinking beyond `!!` and `!$` and maybe `s^`. That's why people only use those. Fish's solution is more interactive, which allows more general replacement of arguments. It is equally or more efficient in terms of number of strokes.

Some users miss `!!` and `!$` and don't want to learn a different way. Should Fish implement just these features as edge cases (see Law of orthogonality)? Or all of Bash history replacement? Or all features from other shells? All features that users can think of?

**Comment by DavidAwad** [comment]
Why does it feel like fish has to shove it's own philosophy unto it's users? isn't the whole point of writing a good shell to give people features they want?

**Comment by DavidAwad** [comment]
Here's a fix for anyone who wants something to copy and paste: 

put this into ~/.config/fish/functions/fish_user_key_bindings.fish

```
# bind control + r to backward search
bind \cr re_search

function bind_bang
  switch (commandline -t)
  case "!"
    commandline -t $history[1]; commandline -f repaint
  case "*"
    commandline -i !
  end
end

function bind_dollar
  switch (commandline -t)
  case "!"
    commandline -t ""
    commandline -f history-token-search-backward
  case "*"
    commandline -i '$'
  end
end

function fish_user_key_bindings
  bind ! bind_bang
  bind '$' bind_dollar
end
```

**Comment by e-beach** [comment]
@siteshwar URL linked to in https://github.com/fish-shell/fish-shell/issues/288#issuecomment-7869918 is broken. Could you edit your comment to fix it for people reading this still-open issue?

**Comment by floam** [comment]
I edited his comment to point to where we host the FAQ now.

**Comment by base698** [comment]
Now that fish has && and || (https://github.com/fish-shell/fish-shell/issues/4620) It's so close to being a full replacement of bash!

**Comment by andysalerno** [comment]
I'm really confused --

So the shell which declares its number 1 design goal is: 

`1. Everything that can be done in other shell languages should be possible to do in fish`

...*doesn't* allow me to do one of the most commonly-used actions in every other shell?

The fish faq says users should instead 

`First press Up, then Home, then type "sudo ".`

Man, what a stretch.  How many users have hit the 'Home' button in the last six months for any reason?  At that length, I could argue that automatically parsing commands from the man page isn't  *orthogonal* enough from just opening up the damn man file and finding it yourself.

**Comment by andreaskern** [comment]
@andysalerno 
the action you want to do is to execute the previous command with sudo prepended
and fish provides this functionality
although instead of entering
`First type "sudo ", then press shift+1, then press shift+1`
I do
`ctrl+p, ctrl+a, type "sudo "`
this works the same way in bash
now that you mentioned the faq I checked it again and somebody removed the part about this shortcut is now missing. (i created https://github.com/fish-shell/fish-shell/pull/3879 to fix this)
The reason this shortcut works is because it follows the `readline` library which is widely implemented or mimicked for line editing.
If I find the time I will create a pull request to add this to the faq
I can understand your frustration, using the `Home` key really isn't an option, as that key is in a different spot on many keyboards (e.g. all of my 3 keyboards).

thx for bringing this up again. I think it is important to have an easy (only modifier and letter keys, no home key , no arrow keys), fast ( not too many keystrokes) and _documented_ way of doing this.

**Comment by andysalerno** [comment]
awesome, I had no idea about ctrl+p!  I like this even better than !!.

**Comment by glmdev** [comment]
It's not as concise, but you can do this with eval and the history variable:

`> man -w fish
/usr/share/fish/man/man1/fish.1

> less (eval $history[1])`

But a better way to do this is using alt-p. Fish will pipe command outputs to less if you press alt-p after you type the command:

`> man fish` **alt-p** `> man fish ^&1 |less;`

**Comment by frosty939** [comment]
i dont understand all of the anti-bang people.. there are so many instances where the bang stuff is laughably faster and more convenient than such "work arounds" as:
`First press Up, then Home, then type "sudo ".`

Bang commands are fucking magic once you get used to them.. I really hope the fish guys change the minds and add them back in. At LEAST a few of the good ones, like:
```
!!
!WORD
!WORD:p
!:n
!$
!^
!n:n
!:gs/MISTAKE/FIX
!#
!#^
```

Fish is pretty nice though, keep up the good work

**Comment by waynew** [comment]
Is there a supported way to execute the last command in history? I don't mind having to type it out, because I'm just using a macro to call the command in my other tmux window. But my usecase is that I just want to be able to do something like this:


    $ echo "hi"
    hi
    $ x
    hi
    $ x
    hi

I tried `eval $history[1]`, but the second call explodes. That's the useful thing about the `!!` command. The one thing that's not really acceptable is using the arrow keys. Is there any command in fish that supports this use case?

**Comment by ridiculousfish** [comment]
The best I can think of right now is to use a leading space which prevents adding it to history:

    $ function x ; eval $history[1] ; end
    $ echo "hi"
    hi
    $  x    # note leading space
    hi
    $  x
    hi

it's certainly not great I know. In this simple case however, the `history-search-backward` binding seems like it should be sufficient? That's arrow keys by default but you can add or change it.

**Comment by waynew** [comment]
Oh! That's a good idea, I didn't think about using the leading space. I'll have to try that.

**Comment by wrecklass** [comment]
This is likely to be the reason I cannot stay with Fish. While it is elegant and simple, it's lack of standard shortcuts for many things makes it non-productive:

First item. I want to repeat the previous command with sudo:
```
sudo !!
```

1) Hit the Up button 
Problem, I have to reach for the Up button on may keyboards. About one-quarter of the time I miss it and then must look down at the keyboard to see exactly where it is and probably fix whatever the miss messed up. Then...
2) Hit Home or Ctrl-A
Home is another stretch and is located in so many diverse locations on different keyboards I will have to look for it. Ctrl-A comes more naturally so that works I guess.
3) Type 'sudo ' and hit return.
Ok, that's only seven keystrokes, but one of them is prone to error and causing my workflow to be interrupted at a large percentage of the time.

Next item, I want to cat the previous last argument:
```
ls -l /etc/network/iptables
cat !$
```
Or with fish:
1) Type 'sudo '
2) Alt-Up
Ok, again this is not very natural. I rarely hit the Alt key and more than often hit the Super (Cmd/Windows/whatever) key next to it which changes the context I'm in altogether.

Fish needs better history expansion like all of the major contenders. Its lack is a reduction in productivity for high end users.

**Comment by nyarly** [comment]
@wrecklass Because I couldn't get `sudo !!` out of my fingers, I wrote https://github.com/nyarly/fish-bang-bang, which I'll suggest to you.

Also, IIRC, fish ships with Alt-S for "prepend sudo" which will work for the last history entry if your command is currently empty, the same way that Alt-P will append `| less`.

**Comment by nyarly** [comment]
And just because it keeps coming up, ("i.e. why doesn't !! work in fish the way it does bash") I want to relate: the special treatment of the `!` character was one of the things that drove me from bash. As a long time fish user, I'd be very disappointed if that behavior were brought over to appease a vocal minority.

**Comment by ClickSentinel** [comment]
`sudo !!` is such a standard thing. I understand that fish is trying very hard to get away from ancient bash syntax, but by pushing away legitimately useful features in favour for personal philosophy, I think we halt progress.

**Comment by tukusejssirs** [comment]
@ClickSentinel, for this very reason I quit using `fish` and stick to `bash`. :smiley:

**Comment by faho** [comment]
Okay, this issue has run its course, so I'm closing it.

Our current plan is to implement global abbreviations, which can then also support "!!" et al. See #5003 for that.

**Comment by scooter-dangle** [comment]
@tukusejssirs and @ClickSentinel, have you tried adding the code at https://github.com/fish-shell/fish-shell/issues/288#issuecomment-306212011 to your config?

I've had a variant of that for several years (modified from an issue (this issue, maybe?) in my config, and it's one of the pieces of shell ergonomics I rely the most on. (Also, especially with the additional case I have in my variant to do an upward token search using the string prior to the `!$`, it's one of the things I'm most annoyed to not have when working in bash.)

My variant (or maybe I stole the variation...can't remember :man_shrugging: ):

```fish
function bind_bang
    switch (commandline --current-token)[-1]
    case "!"
        # Without the `--`, the functionality can break when completing
        # flags used in the history (since, in certain edge cases
        # `commandline` will assume that *it* should try to interpret
        # the flag)
        commandline --current-token -- $history[1]
        commandline --function repaint
    case "*"
        commandline --insert !
    end
end

function bind_dollar
    switch (commandline --current-token)[-1]
    # This case lets us still type a literal `!$` if we need to (by
    # typing `!\$`). Probably overkill.
    case "*!\\"
        # Without the `--`, the functionality can break when completing
        # flags used in the history (since, in certain edge cases
        # `commandline` will assume that *it* should try to interpret
        # the flag)
        commandline --current-token -- (echo -ns (commandline --current-token)[-1] | head -c '-1')
        commandline --insert '$'
    case "!"
        commandline --current-token ""
        commandline --function history-token-search-backward


    # Main difference from referenced version is this `*!` case
    # =========================================================
    #
    # If the `!$` is preceded by any text, search backward for tokens
    # that contain that text as a substring. E.g., if we'd previously
    # run
    #
    #   git checkout -b a_feature_branch
    #   git checkout master
    #
    # then the `fea!$` in the following would be replaced with
    # `a_feature_branch`
    #
    #   git branch -d fea!$
    #
    # and our command line would look like
    #
    #   git branch -d a_feature_branch
    #
    case "*!"
        # Without the `--`, the functionality can break when completing
        # flags used in the history (since, in certain edge cases
        # `commandline` will assume that *it* should try to interpret
        # the flag)
        commandline --current-token -- (echo -ns (commandline --current-token)[-1] | head -c '-1')
        commandline --function history-token-search-backward
    case "*"
        commandline --insert '$'
    end
end

function fish_user_key_bindings
    bind ! bind_bang
    bind '$' bind_dollar
end
```

**Comment by nitantsoni** [comment]
Thanks @scooter-dangle this works even better than [#228](https://github.com/fish-shell/fish-shell/issues/288#issuecomment-306212011)

**Comment by psychoslave** [comment]
There are cases where Fish will provide more idiomatic way to do some action:
```
LANG=C :
fish: Unsupported use of '='. To run ':' with a modified environment, please use 'env LANG=C :…'
```

Maybe `!!` could also, by default, generate such a message suggesting an alternative more idiomatic way to do it?

However the proposal should be just as efficient. Compare:
- `sudo !!<enter>`, all from the base position of hands on the keyboard
- `<up><home>sudo <enter>`, as [suggested in Fish’s FAQ](http://fishshell.com/docs/current/faq.html#why-doesn-t-history-substitution-etc-work) which seems requiring more or less moves of hands depending on where <up> and <home> are on the keyboard.
- `<C-p><Ca>sudo <enter>`, accessible without moving hands , which also works with defaults of Bash, ZSH and maybe more, so users can adopt new habit which is a portable when they have to deal with other shells. 

For `!$`, suggesting `<alt-.>` seems fine, doesn't it?

**Comment by faho** [comment]
@psychoslave Please stop commenting on *all the old issues*, especially closed ones.

Also locking this one. It has been answered about fifty times over.

**Comment by floam** [comment]
I think it was a mistake to mark this as a duplicate and lock conversation: the "global" abbreviation plans don't obviate this issue. Abbreviations will be the mechanism used to implement history expansion but the way this was triaged suggests fish-shell developers specifically decided not to ship the feature or plan on the story being "users should implement it themselves". We can expect a PR implementing history expansion after the new `abbr` lands.

**Comment by nyarly** [comment]
Is the proposal to add a default abbreviation for `!!`, @floam?

**Comment by floam** [comment]
from the [PR](https://github.com/fish-shell/fish-shell/pull/9313)

> Here is the much-longed-for history expansion (https://github.com/fish-shell/fish-shell/issues/288). I think we should ship this, but it would be a separate PR:
> ```fish
> function last_history_item
>     echo $history[1]
> end
> abbr -a !! --position anywhere --function last_history_item --quiet
> ```

**Comment by nyarly** [comment]
Right on. When that lands, I look forward to configuring without `--quiet` which IIUC would be effectively the same as the bind hack I'm using now - replacing `!!` with `history[1]` in place. I've found this 100% better than the Bash behavior.

**Comment by faho** [comment]
Yeah I'm closing this in favor of #9462 - it's possible to add this now, so the question is if we want it by default.

This issue is massive, so it helps to have something a *bit* more focused.
